test/e2e/scale.go

```diff
@@ -0,0 +1,577 @@
+		Expect(os.MkdirAll(input.ArtifactFolder, 0750)).To(Succeed(), "Invalid argument. input.ArtifactFolder can't be created for %s spec", specName)
+
+		Expect(input.E2EConfig.Variables).To(HaveKey(KubernetesVersion))
+
+		// Setup a Namespace where to host objects for this spec and create a watcher for the namespace events.
+		namespace, cancelWatches = setupSpecNamespace(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder)
+	})
+
+	It("Should create a workload cluster", func() {
+		infrastructureProvider := clusterctl.DefaultInfrastructureProvider
+		if input.InfrastructureProvider != nil {
+			infrastructureProvider = *input.InfrastructureProvider
+		}
+
+		flavor := clusterctl.DefaultFlavor
+		if input.Flavor != nil {
+			flavor = *input.Flavor
+		}
+
+		controlPlaneMachineCount := pointer.Int64(1)
+		if input.ControlPlaneMachineCount != nil {
+			controlPlaneMachineCount = input.ControlPlaneMachineCount
+		}
+		// If env variable is defined that will take precedence.
+		if input.E2EConfig.HasVariable(scaleControlPlaneMachineCount) {
+			controlPlaneMachineCountStr := input.E2EConfig.GetVariable(scaleControlPlaneMachineCount)
+			controlPlaneMachineCountInt, err := strconv.Atoi(controlPlaneMachineCountStr)
+			Expect(err).NotTo(HaveOccurred())
+			controlPlaneMachineCount = pointer.Int64(int64(controlPlaneMachineCountInt))
+		}
+
+		workerMachineCount := pointer.Int64(1)
+		if input.WorkerMachineCount != nil {
+			workerMachineCount = input.WorkerMachineCount
+		}
+		// If env variable is defined that will take precedence.
+		if input.E2EConfig.HasVariable(scaleWorkerMachineCount) {
+			workerMachineCountStr := input.E2EConfig.GetVariable(scaleWorkerMachineCount)
+			workerMachineCountInt, err := strconv.Atoi(workerMachineCountStr)
+			Expect(err).NotTo(HaveOccurred())
+			workerMachineCount = pointer.Int64(int64(workerMachineCountInt))
+		}
+
+		clusterCount := input.ClusterCount
+		// If env variable is defined that will take precedence.
+		if input.E2EConfig.HasVariable(scaleClusterCount) {
+			clusterCountStr := input.E2EConfig.GetVariable(scaleClusterCount)
+			var err error
+			clusterCount, err = strconv.Atoi(clusterCountStr)
+			Expect(err).NotTo(HaveOccurred(), "%q value should be integer", scaleClusterCount)
+		}
+
+		concurrency := 1
+		if input.Concurrency > 0 {
+			concurrency = input.Concurrency
+		}
+		// If env variable is defined that will take precedence.
+		if input.E2EConfig.HasVariable(scaleConcurrency) {
+			concurrencyStr := input.E2EConfig.GetVariable(scaleConcurrency)
+			var err error
+			concurrency, err = strconv.Atoi(concurrencyStr)
+			Expect(err).NotTo(HaveOccurred(), "%q value should be integer", scaleConcurrency)
+		}
+
+		// TODO(ykakarap): Follow-up: Add support for legacy cluster templates.
+
+		By("Create the ClusterClass to be used by all workload clusters")
+		// Generate the yaml once and use it as the base template instead of creating the YAML multiple times.
```

Comment:
I'm not sure I got what is the issue about generating and creating the yaml many times.

Reply:
Generating the YAML with the clusterctl lib takes about 1 second => 1000 Clusters => 16 minutes to generate the YAMLs. 

Let's add this to the godoc

Reply:
One alternative to the current implemenattion is to just read & apply the ClusterClass directly from the file and only after that generate the cluster template.

Reply:
I don't think that 1 sec per cluster makes the difference while we are waiting for the cluster to provision before creating the next cluster (also this time should be divide by cuncurrency)

From the other side, this optimization is adding more complexity all around this PR + it makes the resulting flow somehow different from what happens if users scale up.

Reply:
Another thing to consider is that irrespective of if we use `ConfigCluster` to generate the cluster template we always have to install the ClusterClass before we start creating any of the cluster. If we do not, the ClusterClass could be added into the output of many of the template YAML and applying them concurrently causes failures. I observed this failure in the first iteration of this scale test implementation.

Now that since we have to install the ClusterClass first, the alternative would be to install the clusterclass first by doing what we are doing in this PR and then instead of doing a name replacement on a base template to create the target cluster YAML for each cluster we use `ConfigCluster` to create the cluster template for each cluster.

Reply:
We could try to make this safe by adding a mutex. So that no two configcluster's are running at the same time.

Personally I think, in production it's more realistic that folks are just deploying the ClusterClass and then deploying only Cluster objects. But if we can make it work with running ConfigCluster concurrently I'm okay with it.

Reply:
Just slighlty more context. IIRC I was initially also playing around with running ConfigCluster concurrently and I was hitting race conditions. I think it's because ConfigCluster is using singleton's internally. E.g. stuff like this: 
```go
	clusterctllog.SetLogger(log.Logger())
```

Reply:
Discussed with Fabrizio and decided to go with the current approach in the PR. Added more explanation comments.